### PR TITLE
Make the implementation of optimization in DNN adjustable to different vector sizes with RVV intrinsics.

### PR DESCRIPTION
**Pull request by hanliutong**
PR for GSoC'21 project on Optimize OpenCV DNN for RISC-V. Related PR #20287.

This PR will further optimize DNN on the basis of #20287, especially when VLEN>128.

In #20287, we have used RVV Intrinsic to optimize the 4 kernel functions in DNN. However, if RVV vectors are 256-bit wide or more longer, then the current implementation will use only a part of them. This PR tries to adjustable to different vector sizes.

Functions | Implement && Build | block size | Max used of vReg (v0-v31)
-- | -- | -- | --
fastGEMM | ⏱ Ready for review | 4*7 | 32
fastGEMM1T | ⏱ Ready for review | 15*2 | 32
fastConv | ⏱ Ready for review | 3*8  | 28
fastDepthwiseConv | ⏱ Ready for review | / | 18

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ ] I agree to contribute to the project under Apache 2 License.
- [ ] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/dnn/src/layers/layers_common.simd.hpp [hunk 1]
```diff
@@ -1097,23 +1163,27 @@ Example for load_deinterleave:
...
+    vfloat32m4_t tempa = vundefined_f32m4(), tempb = vundefined_f32m4();
+    vfloat32m4_t vw = vle32_v_f32m4(ptr, vl*2);
+    tempa = vcompress_vm_f32m4(mask0, tempa, vw, vl*2);
+    tempb = vcompress_vm_f32m4(mask1, tempb, vw, vl*2);
+    /* The following instructions have not to be supported by the GNU toolchain.
+       So we temporarily use store and load instead.
+    // a = vlmul_trunc_v_f32m4_f32m2(tempa);
+    // b = vlmul_trunc_v_f32m4_f32m2(tempb);
+    */
+    float* buffer = (float*)malloc(sizeof(float32_t)*vl*2);
```
Comment (asmorkalov) [inline modules/dnn/src/layers/layers_common.simd.hpp:1181]: Please use cv::AutoBuffer here: https://docs.opencv.org/master/d8/dd0/classcv_1_1AutoBuffer.html

**Review by asmorkalov — COMMENTED** [review]
:+1: Tested with QEMU and vlen=128,256

File: modules/dnn/src/layers/layers_common.simd.hpp (hunk 1 above)
Comment (hanliutong) [reply modules/dnn/src/layers/layers_common.simd.hpp]: Fixed via 9508aa9

**Review by asmorkalov — APPROVED** [review]
:+1:

**Pull request merged**